File: app/src/main/java/org/secuso/privacyfriendlytodolist/view/dialog/ProcessTodoTaskDialog.java
/*
 This file is part of Privacy Friendly To-Do List.

 Privacy Friendly To-Do List is free software:
 you can redistribute it and/or modify it under the terms of the
 GNU General Public License as published by the Free Software Foundation,
 either version 3 of the License, or any later version.

 Privacy Friendly To-Do List is distributed in the hope
 that it will be useful, but WITHOUT ANY WARRANTY; without even
 the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Privacy Friendly To-Do List. If not, see <http://www.gnu.org/licenses/>.
 */

package org.secuso.privacyfriendlytodolist.view.dialog;

import android.view.ContextMenu;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.fragment.app.FragmentActivity;
import androidx.preference.PreferenceManager;

import org.secuso.privacyfriendlytodolist.R;
import org.secuso.privacyfriendlytodolist.model.Model;
import org.secuso.privacyfriendlytodolist.model.TodoList;
import org.secuso.privacyfriendlytodolist.model.TodoTask;
import org.secuso.privacyfriendlytodolist.util.Helper;

import java.util.List;

/**
 * Created by Sebastian Lutz on 12.03.2018.
 *
 * This class creates a dialog that lets the user create/edit a task.
 */

public class ProcessTodoTaskDialog extends FullScreenDialog<ResultCallback<TodoTask>> {
    private TextView prioritySelector;
    private TextView deadlineTextView;
    private TextView reminderTextView;
    private TextView listSelector;
    private TextView dialogTitleNew;
    private TextView dialogTitleEdit;
    private TextView progressText;
    private TextView progressPercent;
    private RelativeLayout progress_layout;
    private SeekBar progressSelector;
    private EditText taskName;
    private EditText taskDescription;
    private TodoTask.Priority taskPriority = null;
    private Integer selectedListID = null;
    private final List<TodoList> lists;
    private int taskProgress = 0;
    private String name, description;
    private long deadline = -1;
    private long reminderTime = -1;

    private final TodoTask task;

    /**
     * @param context Gets used as context, ViewModelStoreOwner and LifecycleOwner.
     */
    public ProcessTodoTaskDialog(FragmentActivity context, List<TodoList> todoLists) {
        this(context, todoLists, null);
    }

    /**
     * @param context Gets used as context, ViewModelStoreOwner and LifecycleOwner.
     */
    public ProcessTodoTaskDialog(FragmentActivity context, List<TodoList> todoLists, TodoTask todoTask) {
        super(context, R.layout.add_task_dialog);

        lists = todoLists;

        initGui();

        if (null == todoTask){
            task = Model.createNewTodoTask();
            task.setCreated();
        } else {
            task = todoTask;
            task.setChanged();

            deadline = task.getDeadline();
            reminderTime = task.getReminderTime();
            taskName.setText(task.getName());
            taskDescription.setText(task.getDescription());
            prioritySelector.setText(Helper.priority2String(context, task.getPriority()));
            taskPriority = task.getPriority();
            progressSelector.setProgress(task.getProgress(false));
            if (task.getDeadline() <= 0)
                deadlineTextView.setText(context.getString(R.string.no_deadline));
            else
                deadlineTextView.setText(Helper.getDate(deadline));
            if (task.getReminderTime() <= 0)
                reminderTextView.setText(context.getString(R.string.reminder));
            else
                reminderTextView.setText(Helper.getDateTime(reminderTime));
        }
    }

    private void initGui() {

        // initialize textview that displays the selected priority
        prioritySelector = (TextView) findViewById(R.id.tv_new_task_priority);
        prioritySelector.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                registerForContextMenu(prioritySelector);
                openContextMenu(prioritySelector);
            }
        });
        prioritySelector.setOnCreateContextMenuListener(this);
        taskPriority = TodoTask.Priority.DEFAULT_VALUE;
        prioritySelector.setText(Helper.priority2String(getContext(), taskPriority));

        //initialize titles of the dialog
        dialogTitleNew = findViewById(R.id.dialog_title);
        dialogTitleEdit = findViewById(R.id.dialog_edit);


        //initialize textview that displays selected list
        listSelector = findViewById(R.id.tv_new_task_listchoose);
        listSelector.setOnClickListener(v -> {
            registerForContextMenu(listSelector);
            openContextMenu(listSelector);
        });
        listSelector.setOnCreateContextMenuListener(this);


        progressText = findViewById(R.id.tv_task_progress);
        progressPercent = findViewById(R.id.new_task_progress);
        progress_layout = findViewById(R.id.progress_relative);

        // initialize seekbar that allows to select the progress
        progressSelector = (SeekBar) findViewById(R.id.sb_new_task_progress);

        if (!hasAutoProgress()) {
            progressSelector.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {

                @Override
                public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                    taskProgress = progress;
                    progressPercent.setText(progress + "%");
                }

                @Override
                public void onStartTrackingTouch(SeekBar seekBar) {
                }

                @Override
                public void onStopTrackingTouch(SeekBar seekBar) {
                }
            });
        } else {
            makeProgressGone();
        }


        // initialize buttons
        Button okayButton = (Button) findViewById(R.id.bt_new_task_ok);
        okayButton.setOnClickListener(v -> {
            String name = taskName.getText().toString();
            String description = taskDescription.getText().toString();

            if (name.isEmpty()) {
                Toast.makeText(getContext(), getContext().getString(R.string.todo_name_must_not_be_empty), Toast.LENGTH_SHORT).show();
            } /* else if (listName.equals(getContext().getString(R.string.click_to_choose))) {
                Toast.makeText(getContext(), getContext().getString(R.string.to_choose_list), Toast.LENGTH_SHORT).show();
            } */
            else {
                task.setName(name);
                task.setDescription(description);
                task.setDeadline(deadline);
                task.setPriority(taskPriority);
                task.setListId(selectedListID);
                task.setProgress(taskProgress);
                task.setReminderTime(reminderTime);
                callback.onFinish(task);
                ProcessTodoTaskDialog.this.dismiss();
            }
        });

        Button cancelButton = (Button) findViewById(R.id.bt_new_task_cancel);
        cancelButton.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View v) {
                ProcessTodoTaskDialog.this.dismiss();
            }
        });

        // initialize textviews to get deadline and reminder time
        deadlineTextView = (TextView) findViewById(R.id.tv_todo_list_deadline);
        deadlineTextView.setTextColor(okayButton.getCurrentTextColor());
        deadlineTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DeadlineDialog deadlineDialog = new DeadlineDialog(getContext(), deadline);
                deadlineDialog.setDialogCallback(new DeadlineDialog.DeadlineCallback() {
                    @Override
                    public void setDeadline(long d) {
                        deadline = d;
                        deadlineTextView.setText(Helper.getDate(deadline));
                    }

                    @Override
                    public void removeDeadline() {
                        deadline = -1;
                        deadlineTextView.setText(getContext().getResources().getString(R.string.deadline));
                    }
                });
                deadlineDialog.show();
            }
        });

        reminderTextView = (TextView) findViewById(R.id.tv_todo_list_reminder);
        reminderTextView.setTextColor(okayButton.getCurrentTextColor());
        reminderTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ReminderDialog reminderDialog = new ReminderDialog(getContext(), reminderTime, deadline);
                reminderDialog.setDialogCallback(new ReminderDialog.ReminderCallback() {
                    @Override
                    public void setReminder(long r) {

                        /*if(deadline == -1) {
                            Toast.makeText(getContext(), getContext().getString(R.string.set_deadline_before_reminder), Toast.LENGTH_SHORT).show();
                            return;
                        }*/

                        if(deadline != -1 && deadline < r) {
                            Toast.makeText(getContext(), getContext().getString(R.string.deadline_smaller_reminder), Toast.LENGTH_SHORT).show();
                            return;
                        }

                        reminderTime = r;
                        reminderTextView.setText(Helper.getDateTime(reminderTime));
                    }

                    @Override
                    public void removeReminder() {
                        reminderTime = -1;
                        TextView reminderTextView = (TextView) findViewById(R.id.tv_todo_list_reminder);
                        reminderTextView.setText(getContext().getResources().getString(R.string.reminder));
                    }
                });
                reminderDialog.show();
            }
        });

        taskName = (EditText) findViewById(R.id.et_new_task_name);
        taskDescription = (EditText) findViewById(R.id.et_new_task_description);
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
        switch (v.getId()){
            case R.id.tv_new_task_priority:
                menu.setHeaderTitle(R.string.select_priority);
                for (TodoTask.Priority priority : TodoTask.Priority.getEntries()) {
                    menu.add(Menu.NONE, priority.ordinal(), Menu.NONE, Helper.priority2String(getContext(), priority));
                }
                break;

            case R.id.tv_new_task_listchoose:
                menu.setHeaderTitle(R.string.select_list);
                menu.add(Menu.NONE, -1, Menu.NONE, R.string.select_no_list);
                for (int i = 0; i < lists.size(); ++i) {
                    TodoList todoList = lists.get(i);
                    // Add offset so that IDs are non-overlapping with priority-IDs
                    menu.add(Menu.NONE, TodoTask.Priority.LENGTH + i, Menu.NONE, todoList.getName());
                }
            break;
        }
    }


    @Override
    public boolean onMenuItemSelected(int featureId, MenuItem item) {
        if (item.getItemId() >= 0 && item.getItemId() < TodoTask.Priority.LENGTH) {
            taskPriority = TodoTask.Priority.getEntries().get(item.getItemId());
            prioritySelector.setText(Helper.priority2String(getContext(), taskPriority));
        } else {
            final int todoListIndex = item.getItemId() - TodoTask.Priority.LENGTH;
            TodoList todoList = null;
            if (todoListIndex >= 0 && todoListIndex < lists.size()) {
                todoList = lists.get(todoListIndex);
            }
            setListSelector(todoList);
        }

        return super.onMenuItemSelected(featureId, item);
    }

    //change the dialog title from "new task" to "edit task"
    public void titleEdit(){
        dialogTitleNew.setVisibility(View.GONE);
        dialogTitleEdit.setVisibility(View.VISIBLE);
    }

    //sets the textview either to list name in context or if no context to default
    public void setListSelector(Integer todoListId) {
        TodoList todoList = null;
        if (null != todoListId) {
            for (TodoList currentTodoList : lists) {
                if (currentTodoList.getId() == todoListId) {
                    todoList = currentTodoList;
                    break;
                }
            }
        }
        setListSelector(todoList);
    }

    private void setListSelector(TodoList todoList) {
        if (null != todoList) {
            selectedListID = todoList.getId();
            listSelector.setText(todoList.getName());
        } else {
            selectedListID = null;
            listSelector.setText(getContext().getString(R.string.click_to_choose));
        }
    }

    private boolean hasAutoProgress() {
        //automatic-progress enabled?
        if (!PreferenceManager.getDefaultSharedPreferences(getContext()).getBoolean("pref_progress", false))
            return false;
        return true;
    }

    //Make progress-selectionbar disappear
    private void makeProgressGone() {
        progress_layout.setVisibility(View.GONE);
           /* progressSelector.setVisibility(View.INVISIBLE);
            progressPercent.setVisibility(View.INVISIBLE);
            progressText.setVisibility(View.INVISIBLE); */
    }
}

File: app/src/main/java/org/secuso/privacyfriendlytodolist/view/dialog/ProcessTodoTaskDialog.kt
/*
Privacy Friendly To-Do List
Copyright (C) 2018-2025  Sebastian Lutz

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package org.secuso.privacyfriendlytodolist.view.dialog

import android.os.Bundle
import android.util.Log
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import android.widget.Button
import android.widget.EditText
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.FragmentActivity
import androidx.preference.PreferenceManager
import org.secuso.privacyfriendlytodolist.R
import org.secuso.privacyfriendlytodolist.model.Model
import org.secuso.privacyfriendlytodolist.model.TodoTask
import org.secuso.privacyfriendlytodolist.model.TodoTask.RecurrencePattern
import org.secuso.privacyfriendlytodolist.util.Helper
import org.secuso.privacyfriendlytodolist.util.LogTag
import org.secuso.privacyfriendlytodolist.util.PreferenceMgr
import org.secuso.privacyfriendlytodolist.util.Timestamp
import org.secuso.privacyfriendlytodolist.viewmodel.CustomViewModel
import java.util.Locale

/**
 * This class creates a dialog that lets the user create/edit a task.
 *
 * Created by Sebastian Lutz on 12.03.2018.
 */
@Suppress("UNUSED_ANONYMOUS_PARAMETER")
class ProcessTodoTaskDialog(context: FragmentActivity,
                            private var todoTask: TodoTask):
        FullScreenDialog<ResultCallback<TodoTask>>(context, R.layout.task_dialog) {

    private var deadline: Timestamp?
    private var recurrencePattern: RecurrencePattern
    private var recurrenceInterval: Int
    private var reminderTime: Timestamp?
    private var taskProgress: Int
    private var taskPriority: TodoTask.Priority
    private var assignedTodoListId: Int?

    private var todoLists: Map<Int, String> = mapOf()
    private var editExistingTask: Boolean = true

    // GUI elements
    private lateinit var recurrencePatternTextView: TextView
    private lateinit var recurrenceIntervalEditText: EditText
    private lateinit var progressPercent: TextView
    private lateinit var prioritySelector: TextView
    private lateinit var listSelector: TextView

    private enum class GroupId {
        TASK_PRIORITY, NO_TASK_LIST, TASK_LIST_CHOOSE
    }

    init {
        deadline = todoTask.getDeadline()
        recurrencePattern = todoTask.getRecurrencePattern()
        recurrenceInterval = todoTask.getRecurrenceInterval()
        reminderTime = todoTask.getReminderTime()
        taskProgress = todoTask.getProgress()
        taskPriority = todoTask.getPriority()
        assignedTodoListId = todoTask.getListId()
    }

    constructor(context: FragmentActivity,
                todoListId: Int?) : this(context, Model.createNewTodoTask()) {
        assignedTodoListId = todoListId
        editExistingTask = false
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Dialog title
        if (editExistingTask) {
            findViewById<Toolbar>(R.id.task_dialog_title).setTitle(R.string.edit_todo_task)
        }
        // Task name
        val taskName: EditText = findViewById(R.id.et_task_name)
        taskName.setText(todoTask.getName())
        // Task description
        val taskDescription: EditText = findViewById(R.id.et_task_description)
        taskDescription.setText(todoTask.getDescription())
        // Task deadline
        val deadlineTextView: TextView = findViewById(R.id.tv_todo_list_deadline)
        deadlineTextView.text = deadline?.createLocalizedDateString() ?: context.getString(R.string.deadline)
        deadlineTextView.setOnClickListener {
            val deadlineDialog = DeadlineDialog(context, deadline)
            deadlineDialog.setDialogCallback(object : DeadlineCallback {
                override fun setDeadline(selectedDeadline: Timestamp) {
                    deadline = selectedDeadline
                    deadlineTextView.text = selectedDeadline.createLocalizedDateString()
                }

                override fun removeDeadline() {
                    deadline = null
                    deadlineTextView.text = context.resources.getString(R.string.deadline)
                }
            })
            deadlineDialog.show()
        }
        // Task reminder
        val reminderTextView: TextView = findViewById(R.id.tv_todo_list_reminder)
        reminderTextView.text = reminderTime?.createLocalizedDateTimeString() ?: context.getString(R.string.reminder)
        reminderTextView.setOnClickListener {
            val reminderDialog = ReminderDialog(context, reminderTime, deadline)
            reminderDialog.setDialogCallback(object : ReminderCallback {
                override fun setReminderTime(selectedReminderTime: Timestamp) {
                    var resIdErrorMsg = 0
                    val deadlineCopy = deadline
                    if (recurrencePattern == RecurrencePattern.NONE) {
                        /* if (deadline == null) {
                            resIdErrorMsg = R.string.set_deadline_before_reminder
                        } else */
                        if (deadlineCopy != null && deadlineCopy < selectedReminderTime) {
                            resIdErrorMsg = R.string.deadline_smaller_reminder
                        } else if (selectedReminderTime < Timestamp.createCurrent()) {
                            resIdErrorMsg = R.string.reminder_smaller_now
                        }
                    }
                    if (resIdErrorMsg == 0) {
                        reminderTime = selectedReminderTime
                        reminderTextView.text = selectedReminderTime.createLocalizedDateTimeString()
                    } else {
                        Toast.makeText(context, context.getString(resIdErrorMsg), Toast.LENGTH_SHORT).show()
                    }
                }

                override fun removeReminderTime() {
                    reminderTime = null
                    reminderTextView.text = context.resources.getString(R.string.reminder)
                }
            })
            reminderDialog.show()
        }
        // Task recurrence pattern
        recurrencePatternTextView = findViewById(R.id.tv_task_recurrence_pattern)
        recurrencePatternTextView.setOnClickListener {
            val dialog = RecurrencePatternDialog(context, recurrencePattern)
            dialog.setDialogCallback { newRecurrencePattern ->
                recurrencePattern = newRecurrencePattern
                updateRecurrencePatternText()
            }
            dialog.show()
        }
        recurrencePatternTextView.setOnCreateContextMenuListener(this)
        recurrenceIntervalEditText = findViewById(R.id.tv_task_recurrence_interval)
        updateRecurrencePatternText()
        // Task progress
        val progressSelector: SeekBar = findViewById(R.id.sb_task_progress)
        progressPercent = findViewById(R.id.tv_task_progress)
        progressSelector.progress = taskProgress
        updateProgressPercentText()
        if (hasAutoProgress()) {
            findViewById<TextView>(R.id.tv_task_progress_str).visibility = View.GONE
            progressSelector.visibility = View.GONE
            progressPercent.visibility = View.GONE
        } else {
            progressSelector.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                    taskProgress = progress
                    updateProgressPercentText()
                }

                override fun onStartTrackingTouch(seekBar: SeekBar) {}

                override fun onStopTrackingTouch(seekBar: SeekBar) {}
            })
        }
        // Task priority
        prioritySelector = findViewById(R.id.tv_task_priority)
        prioritySelector.text = Helper.priorityToString(context, taskPriority)
        prioritySelector.setOnCreateContextMenuListener(this)
        prioritySelector.setOnClickListener {
            registerForContextMenu(prioritySelector)
            openContextMenu(prioritySelector)
        }
        // Task list
        listSelector = findViewById(R.id.tv_task_list_choose)
        listSelector.setOnClickListener {
            registerForContextMenu(listSelector)
            openContextMenu(listSelector)
        }
        listSelector.setOnCreateContextMenuListener(this)
        // OK button
        val okayButton: Button = findViewById(R.id.bt_process_task_ok)
        okayButton.setOnClickListener {
            val name = taskName.text.toString()
            val description = taskDescription.text.toString()
            val recurrenceIntervalText = recurrenceIntervalEditText.text.toString()
            val recurrenceInterval = recurrenceIntervalText.toIntOrNull()
            val isRecurrenceIntervalBad = recurrenceInterval == null || recurrenceInterval < 1
            if (name.isEmpty()) {
                Toast.makeText(context, context.getString(R.string.todo_name_must_not_be_empty),
                    Toast.LENGTH_SHORT).show()
            } else if (recurrencePattern != RecurrencePattern.NONE && deadline == null) {
                Toast.makeText(context, context.getString(R.string.set_deadline_if_recurring),
                    Toast.LENGTH_SHORT).show()
            } else if (recurrencePattern != RecurrencePattern.NONE && isRecurrenceIntervalBad) {
                Toast.makeText(context, context.getString(R.string.recurrence_interval_invalid),
                    Toast.LENGTH_SHORT).show()
            } else {
                todoTask.setName(name)
                todoTask.setDescription(description)
                todoTask.setDeadline(deadline)
                todoTask.setRecurrencePattern(recurrencePattern)
                if (recurrencePattern in RecurrencePattern.WEEKDAYS_M______ .. RecurrencePattern.WEEKDAYS_MTWTFSS) {
                    // For these pattern an interval of 1 gets forced.
                    todoTask.setRecurrenceInterval(1)
                } else if ( ! isRecurrenceIntervalBad) {
                    todoTask.setRecurrenceInterval(recurrenceInterval)
                }
                todoTask.setReminderTime(reminderTime)
                todoTask.setProgress(taskProgress)
                todoTask.setPriority(taskPriority)
                todoTask.setListId(assignedTodoListId)
                todoTask.setChanged()
                getDialogCallback().onFinish(todoTask)
                dismiss()
            }
        }
        // Cancel button
        val cancelButton: Button = findViewById(R.id.bt_process_task_cancel)
        cancelButton.setOnClickListener { dismiss() }

        val viewModel = CustomViewModel(context)
        val model = viewModel.model
        model.getAllToDoListNames { allToDoListNames ->
            todoLists = allToDoListNames
            updateListSelector()
        }

        if (!editExistingTask) {
            // Request focus for first input field.
            taskName.requestFocus()
            // Show soft-keyboard
            window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
        }
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenuInfo?) {
        when (v.id) {
            R.id.tv_task_priority -> {
                val menuHeader = Helper.getMenuHeader(layoutInflater, v, R.string.select_priority)
                menu.setHeaderView(menuHeader)
                for (priority in TodoTask.Priority.entries) {
                    menu.add(GroupId.TASK_PRIORITY.ordinal, priority.ordinal, Menu.NONE,
                        Helper.priorityToString(context, priority))
                }
            }

            R.id.tv_task_list_choose -> {
                val menuHeader = Helper.getMenuHeader(layoutInflater, v, R.string.select_list)
                menu.setHeaderView(menuHeader)
                menu.add(GroupId.NO_TASK_LIST.ordinal, Menu.NONE, Menu.NONE, R.string.select_no_list)
                for (entry in todoLists.entries) {
                    menu.add(GroupId.TASK_LIST_CHOOSE.ordinal, entry.key, Menu.NONE, entry.value)
                }
            }
        }
    }

    override fun onMenuItemSelected(featureId: Int, item: MenuItem): Boolean {
        when (item.groupId) {
            GroupId.TASK_PRIORITY.ordinal -> {
                taskPriority = TodoTask.Priority.fromOrdinal(item.itemId)!!
                prioritySelector.text = Helper.priorityToString(context, taskPriority)
            }

            GroupId.NO_TASK_LIST.ordinal -> {
                assignedTodoListId = null
                updateListSelector()
            }

            GroupId.TASK_LIST_CHOOSE.ordinal -> {
                assignedTodoListId = item.itemId
                updateListSelector()
            }

            else -> {
                Log.e(TAG, "Unhandled menu item group ID ${item.groupId}.")
            }
        }

        return super.onMenuItemSelected(featureId, item)
    }

    private fun hasAutoProgress(): Boolean {
        //automatic-progress enabled?
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getBoolean(PreferenceMgr.P_IS_AUTO_PROGRESS.name, false)
    }

    private fun updateProgressPercentText() {
        val text = "$taskProgress %"
        progressPercent.text = text
    }

    private fun updateRecurrencePatternText() {
        if (recurrencePattern == RecurrencePattern.NONE) {
            recurrencePatternTextView.text = context.getString(R.string.recurrence_pattern_hint)
            recurrenceIntervalEditText.setText("")
        } else {
            recurrencePatternTextView.text = Helper.recurrencePatternToNounString(context, recurrencePattern)
            val forceInterval1 = (recurrencePattern in RecurrencePattern.WEEKDAYS_M______ .. RecurrencePattern.WEEKDAYS_MTWTFSS)
            if (forceInterval1) {
                recurrenceInterval = 1
            }
            if (forceInterval1 || recurrenceIntervalEditText.text.isEmpty()) {
                recurrenceIntervalEditText.setText(String.format(Locale.getDefault(), "%d", recurrenceInterval))
            }
        }
    }

    private fun updateListSelector() {
        var text: String? = null
        if (null != assignedTodoListId) {
            text = todoLists[assignedTodoListId]
        }
        if (null == text) {
            assignedTodoListId = null
            text = context.getString(R.string.click_to_choose)
        }
        listSelector.text = text
    }

    companion object {
        private val TAG = LogTag.create(this::class.java.declaringClass)
    }
}
